ruma-events-macros/src/collection.rs

```diff
@@ -0,0 +1,65 @@
+//! Implementation of the collection type macro.
+
+use proc_macro2::{TokenStream};
+use quote::quote;
+use syn::{
+    parse_quote, parse_str
+};
+
+use crate::{
+    collection_parse::RumaCollectionInput,
+    gen::to_camel_case,
+};
+
+/// Create a collection from `RumaCollectionInput.
+pub fn expand_collection(input: RumaCollectionInput) -> syn::Result<TokenStream> {
+    let attrs = &input.attrs;
+    let ident = &input.name;
+    let mut variants = vec![];
+    for (idx, var) in input
+        .events
+        .iter()
+        .map(|lit| {
+            format!(
+                "{}({})",
+                to_camel_case(lit.value()),
+                to_event_content(lit.value())
+            )
+        })
+        .map(|s| parse_str::<syn::Variant>(&s))
+        .enumerate()
+    {
+        let mut var = var?;
+
+        let content_docstring = &input.events[idx];
+        var.attrs.push(parse_quote! { #[doc = #content_docstring] });
+        
+        variants.push(var)
+    }
+
+    let collection = quote! {
+        #( #attrs )*
+        #[derive(Clone, Debug, Serialize)]
+        #[serde(untagged)]
+        #[allow(clippy::large_enum_variant)]
+        pub enum #ident {
+            #( #variants ),*
+        }
+    };
+
+    Ok(collection)
+}
+
+/// Splits the given `event_type` string on `.` and `_` removing the `m.` then
+/// using only the event name append "EventContent".
+fn to_event_content(name: String) -> String {
```

Comment:
Same as `to_camel_case`:

```suggestion
fn to_event_content(name: LitStr) -> Ident {
```